src/riak_cs_s3_auth.erl

```diff
@@ -130,43 +172,157 @@ calculate_signature(KeyData, RD) ->
+    authenticate_v4(SecretAccessKey, AuthAttrs, Method, Path, Qs, AllHeaders).
+
+authenticate_v4(SecretAccessKey, AuthAttrs, Method, Path, Qs, AllHeaders) ->
+    CanonicalRequest = canonical_request_v4(AuthAttrs, Method, Path, Qs, AllHeaders),
+    _ = lager:debug("CanonicalRequest(v4):~n~s~nEND-OF-CANONICAL-REQUEST", [CanonicalRequest]),
+    {StringToSign, Scope} =
+        string_to_sign_v4(AuthAttrs, AllHeaders, CanonicalRequest),
+    _ = lager:debug("StringToSign(v4):~n~s~nEND-OF-STRING-TO-SIGN", [StringToSign]),
+    CalculatedSignature = calculate_signature_v4(SecretAccessKey, Scope, StringToSign),
+    _ = lager:debug("CalculatedSignature(v4): ~s", [CalculatedSignature]),
+    {"Signature", PresentedSignature} = lists:keyfind("Signature", 1, AuthAttrs),
+    _ = lager:debug("PresentedSignature(v4)  : ~s", [PresentedSignature]),
+    case CalculatedSignature of
+        PresentedSignature -> ok;
+        _ -> {error, {unmatched_signature, PresentedSignature, CalculatedSignature}}
+    end.
+
+canonical_request_v4(AuthAttrs, Method, Path, Qs, AllHeaders) ->
+    CanonicalQs = canonicalize_qs(v4, Qs),
+    {"SignedHeaders", SignedHeaders} = lists:keyfind("SignedHeaders", 1, AuthAttrs),
+    CanonicalHeaders = canonical_headers_v4(AllHeaders,
+                                            string:tokens(SignedHeaders, [$;])),
+    {"x-amz-content-sha256", HashedPayload} =
+        lists:keyfind("x-amz-content-sha256", 1, AllHeaders),
+    CanonicalRequest = [atom_to_list(Method), $\n,
+                        strict_url_encode_for_path(Path), $\n,
+                        CanonicalQs, $\n,
+                        CanonicalHeaders, $\n,
+                        SignedHeaders, $\n,
+                        HashedPayload],
+    CanonicalRequest.
+
+canonical_headers_v4(AllHeaders, HeaderNames) ->
+    %% TODO:
+    %% - Assert `host' and all `x-amz-*' in AllHeaders are included in `HeaderNames'
+    %% - Assert `conetnt-type' is included in `HeaderNames' if it is included in
+    %%   `AllHeaders'
+    canonical_headers_v4(AllHeaders, HeaderNames, []).
+
+canonical_headers_v4(_AllHeaders, [], Acc) ->
+    Acc;
+canonical_headers_v4(AllHeaders, [HeaderName | Rest], Acc) ->
+    {HeaderName, Value} = lists:keyfind(HeaderName, 1, AllHeaders),
+    canonical_headers_v4(AllHeaders, Rest, [Acc | [HeaderName, $:, Value, $\n]]).
+
+string_to_sign_v4(AuthAttrs, AllHeaders, CanonicalRequest) ->
+    %% TODO: Expires?
+    TimeStamp = case lists:keyfind("x-amz-date", 1, AllHeaders) of
+                    false ->
+                        {"date", Date} = lists:keyfind("date", 1, AllHeaders),
+                        riak_cs_wm_utils:to_iso_8601(Date);
+                    {"x-amz-date", XAmzDate} ->
+                        XAmzDate
+                end,
+    {"Credential", Cred} = lists:keyfind("Credential", 1, AuthAttrs),
+    [_UserId, CredDate, AwsRegion, "s3" = AwsService, "aws4_request" = AwsRequest] =
```

Comment:
Yeah, it does such workaround.

So far, validating region does not improve security [1] and introduces some complexity for clients.
I will add comments but does not change code.

[1] For AWS, signing keys can be created by super-admin and normally one can use signing keys to access S3 and does not need to use his/her original-only-one secret key usually. Is it also nice for Riak CS and public service by CS? Hmm, maybe.